buildscripts/covertest.py

```diff
@@ -6,79 +6,177 @@
+class CoverageRunner(object):
     """
-    Run go test with coverage over the the given packages, with the following options
+    CoverageRunner generates coverage profiles for a packages within a base package,
+    making sure to produce numbers for every relevant package that it depends on or that its
+    tests depends on.
+
+    Ideally, we can just do `-coverpkg=all`, but (1) that includes all vendored packages,
+    and (2) even if we do `-coverpkg=$(go list ./... | grep -v vendor)`, compiling
+    the test binary slows down a bit for each package in the `-coverpkg` list.
+
+    This has to calculate the recursive dependences for the package being tested, which
+    is the union of the following sets of dependencies:gi
+    - recursive dependencies of the non-test package: produced with `go list -f {{.Deps}}`
+    - test imports (which are not necessarily included with the previous go list command):
+      produced with `go list -f {{.TestImports}}`
+    - test imports from a test in a different package, such as from package `<pkg>_test`:
+      produced with `go list -f {{.XTestImports}}`
     """
-    all_pkgs = get_all_notary_pkgs(buildtags)
-    all_pkgs.sort()
-
-    pkgs = pkgs or all_pkgs
-    allpkg_string = ",".join(all_pkgs)
-
-    base_cmd = ["go", "test", "-covermode", covermode, "-coverpkg", allpkg_string] + opts.split()
-    if buildtags:
-        base_cmd.extend(["-tags", buildtags])
-
-    allpkg_string = ", ".join(all_pkgs)
-    longest_pkg_len = max([len(pkg) for pkg in pkgs])
-
-    for pkg in pkgs:
-        cmd = base_cmd + ["-coverprofile", os.path.join(coverdir, get_coverprofile_filename(pkg, buildtags)), pkg]
-        app = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
-        for line in app.stdout:
-            # we are trying to generate coverage reports for everything in the base package, and some may not be
-            # actually exercised in the test.  So ignore this particular warning.  Also, we want to pretty-print
-            # the test success/failure results
-            if not line.startswith("warning: no packages being tested depend on github.com/docker/notary"):
-                line = line.replace(allpkg_string, "<all packages>").replace(pkg, pkg.ljust(longest_pkg_len))
-                sys.stdout.write(line)
-
-        app.wait()
-        if app.returncode != 0:
-            print("\n\nTests failed.\n")
-            sys.exit(app.returncode)
+    def __init__(self, buildtags):
+        self.base_pkg = subprocess.check_output("go list".split()).strip()
+        self.buildtags = buildtags
+        self.tag_args = ()
+        if buildtags != "":
+            self.tag_args = ("-tags", self.buildtags)
+
+        self.recursive_pkg_deps, self.test_imports = self._get_all_pkg_info()
+
+    def _filter_pkgs(self, pkgs):
+        """
+        Given a list of packages, returns list that contains only the packages
```

Comment:
this comment reads funny.  Maybe `...returns a filtered copy of the list that only contains source packages derived from the base package`?